File: src/alfred.php
<?php

require_once __DIR__ . '/noteplan.php';

function alfred_itemize(array $items): string {
    return json_encode(['items' => $items]);
}

function alfred_return_error(Throwable $th): string {
    return alfred_itemize([
        ['title' => 'Error: ' . $th->getMessage(), 'arg' => $th->getMessage(), 'valid' => false]
    ]);
}

function alfred_query_to_sqlite(array $argv): array {
    // $a[0] is the script name
    $originalQuery = $argv[1]; 

    // Remove everything except numbers, letters and spaces
    $sqliteQuery = preg_replace('/[^\p{L}\s\d]/', '', $originalQuery);
    // compress spaces
    $sqliteQuery = preg_replace('/\s+/', ' ', $sqliteQuery);
    // modify each word => word*
    $sqliteQuery = str_replace(' ', '* ', $sqliteQuery). '*';

    return [
        $originalQuery,
        $sqliteQuery
    ];
}

function alfred_query_to_rg_regex(array $argv): array {
    // $argv[0] is the script name
    $rawQuery = $argv[1]; 

    // Remove everything except numbers, letters and spaces
    $rawQuery = preg_replace('/[^\p{L}\s\d]/', '', $rawQuery);
    // compress spaces
    $rawQuery = preg_replace('/\s+/', ' ', $rawQuery);

    $lookupParts = explode(' ', $rawQuery);
    $last = array_pop($lookupParts);

    // append suffix to each element: explode -> map -> implode 
    return [
        $rawQuery,
        implode('', [...array_map(fn($l) => "{$l}(?s:(?!{$l}).)*?", $lookupParts), $last]),
    ];
}

function alfred_create_note_item(string $title, array $config) {
    return [
        'title' => "Create '$title'",
        'subtitle' => 'Create a new note',
        'arg' => $title,
        'icon' => [
            'path' => __DIR__ . "/../icons/icon-create.icns",
        ]
    ];
}

File: lib/alfred.dart
/*
 * Helper functions for Alfred
 *
 * - alf_to_results
 * - alf_invalid_item
 * - alf_valid_item
 */

import 'dart:convert';
import 'dart:io';

import 'package:alfred_noteplan/strings.dart';

String alf_to_results (List<Map> results) {
	return jsonEncode({'items': results});
}

void alf_exit(List<Map> results) {
	print(alf_to_results(results));
	exit(0);
}

Map<String, dynamic> alf_item (
	String title,
	String subtitle,
	{
		String? uid,
		Map? icon,
		String? arg,
		Map? mods,
		Map? variables,
		Map? text,
		String? quicklookurl,
		bool valid = true
	}
) {
	final Map<String, dynamic> result = {
		'title': title,
		'subtitle': subtitle,
		'valid': valid
	};
	if (uid != null)  { result['uid'] = uid; }
	if (icon != null) { result['icon'] = icon; }
	if (arg != null)  { result['arg']  = arg;}
	if (mods != null) { result['mods'] = mods; }
	if (text != null) { result['text'] = text; }
	if (variables != null) { result['variables'] = variables; }
	if (quicklookurl != null) { result['quicklookurl'] = quicklookurl; }
	return result;
}

Map<String, dynamic> alf_create_item(String query) => alf_item(
	'Create "${query}"',
	str_create_result_subtitle,
	icon: {'path': 'icons/icon-create.icns'},
	arg: query,
	variables: {'action': 'create'}
);